### ISSUE TITLE
Questions

### ISSUE THREAD

**Issue by Avani8** [lowRISC/opentitan#1285]
1) how to increase ROM and flash sizes? 
based on my understanding .. I tried changing the size 
```
memory: [
    { name: "rom",
      clock_connections: {clk_i: "main"},
      reset_connections: {rst_ni: "sys"},
      type: "rom",
      base_addr: "0x00008000",
      size: "0x2000"
    },
    { name: "ram_main",
      clock_connections: {clk_i: "main"},
      reset_connections: {rst_ni: "sys"},
      type: "ram_1p",
      base_addr: "0x10000000",
      size: "0x10000" },
    { name: "eflash",
      clock_connections: {clk_i: "main"},
      reset_connections: {rst_ni: "lc"},
      type: "eflash",
      base_addr: "0x20000000",
      size: "0x80000" },
  ],
```

in hw/top_earlgrey/data/top_earlgrey.hjson. however, I don't know how can I verify the change. please advise. 

2)  if I want to add second stage bootloader  (please excuse me if my question is so basic) what I am doing is providing next image location in sw/device/boot_rom/rom_link.ld 
```
like its pointing to 
_flash_start = ORIGIN(flash);
_flash_start2 = ORIGIN(flash) + 0x100;


```
and calling _flash_start2 in hello_world.c example  as its done in boot_rom.c asm code 
```
static noreturn void try_launch(void) {
  asm volatile(
      "  la   sp, _stack_start;"
      "  tail _flash_start2;"
      // Unfortunately, we can't tell the compiler that we're about to clobber
      // the stack pointer (which is meaningless, since there's nowhere to spill
      // sp to). However, this doesn't matter, since this asm block will never
      // return.
      :
      :
      : /* sp, */ "memory");
  __builtin_unreachable();
}
```
am i right or missing somethings?

**Comment by tjaychen** [comment]
thanks for pointing this out!
Sorry the answer here is going to be a bit long and is actually a slight mistake on our end.

For rom / ram, you do indeed change in the hjson, however, that is not immediately reflected until you generate the final output.  To do so, go to `hw` directory, and just do a `make all`.  Then you'll see a bunch of git diffs in your area.

Flash right now seems to not follow this process, and have its overall sized coded in `hw/top_earlgrey/rtl/top_pkg.sv`.  This is something that needs to be fixed to be more inline with the other 2 and I will file an issue for this.

For the boot loaders, @mcy can give a better answer.

**Comment by eunchan** [comment]
Also please remind that the master branch has a change of hjson format. `base_addr` and `size_byte` are now changed to `addr_range` field.

**Comment by Avani8** [comment]
@tjaychen  thanks for response. I tried changing ROM size from 2000 to 6000 as we per our need and added custom c code in bootstrap.c.  I did following steps
```
$ cd  hw
$ make clean 
$ make all
$ cd ..
$ ninja -C build-out/sw/fpga sw/device/boot_rom/boot_rom_export
```
it compiles fine but after that when I am running 
```
$  fusesoc --cores-root . run --target=synth lowrisc:systems:top_earlgrey_nexysvideo
```
it fails with error 
```
ERROR: Failed to build lowrisc:systems:top_earlgrey_nexysvideo:0.1 : 'make' exited with an error code
`
based on the rumme.log ( see the attachments) . I checked xbar_main.sv and it looks like it is failing at ADDR_SPACE_ROM  or where the address lines are defined ...I believe this is generated using autogen tool but I am not sure where else it is having hardcoded value for rom size. 


![Screenshot from 2020-01-09 02-19-57](https://user-images.githubusercontent.com/12041484/72046523-97158d80-3286-11ea-9644-fe10318c3b8a.png)
![Screenshot from 2020-01-09 02-18-05](https://user-images.githubusercontent.com/12041484/72046524-97158d80-3286-11ea-9562-36065be365de.png)
![Screenshot from 2020-01-09 02-15-48](https://user-images.githubusercontent.com/12041484/72046525-97158d80-3286-11ea-9dd6-6aef6812eba2.png)

**Comment by tjaychen** [comment]
thanks for the report! we will have a look at this.

**Comment by tjaychen** [comment]
@eunchan for visibility

**Comment by tjaychen** [comment]
hi! we think we fixed this in #1309 .. could you give that an update and try again?
I suspect you still might run into FPGA issues because of the way some of post processing scripts work..but give it a try first.

At least there should be no syntax errors.

**Comment by Avani8** [comment]
@tjaychen  thanks for fixing this...now it passes the build and I can deploy it on FPGA but it is hanging on frame 1 check.

![Screenshot from 2020-01-10 14-05-14](https://user-images.githubusercontent.com/12041484/72179082-456a2180-33b2-11ea-8dd5-cd58cba4d4d0.png)
![Screenshot from 2020-01-10 14-04-53](https://user-images.githubusercontent.com/12041484/72179083-456a2180-33b2-11ea-864d-a172252c7025.png)

as far as I follow...bootstrap takes 1kb chunks of the executable image, hashes it and compares the hashes. If the hash matches what the host expects, the host then continues to send the next frame; otherwise, it retries the previous frame.

now the question is how pre-computed hash is created to compare with? how does one modify it? 
in other words, what I am seeing is for first frame hashes are matching with the golden so it goes to the next step but for the next frame, it doesn't match and it is stuck

**Comment by tjaychen** [comment]
@Avani8 
your understanding is correct.
You do not need to change the pre-computed hash, it is actually done as part of the spiflash program, see [here](https://github.com/lowRISC/opentitan/blob/master/sw/host/spiflash/updater.cc#L128) 

At this point we'll probably want to do some debugging to figure out what the ROM is doing because it seems to not be computing the right output.  Before we do that though, can you confirm on your system if you used the existing ROM size (0x2000), everything works fine (as in you can flash in a new program)?

If yes, our next step, we'll probably want to add some printouts inside the bootstrap function to see what it's doing.

@mcy 
@moidx

**Comment by Avani8** [comment]
hi @tjaychen ..yes with 0x2000 everything works fine...to give some more insight I've created sample sum, sub, mult, div functions in c and added that file under sw/device/lib/ included it in meson build and added some uart print results ...that program too works fine. 
the problem comes, when I add a big source file(example compression function) and increase the ROM size.

**Comment by Avani8** [comment]
now to give details, what I am trying to do here is add data compression logic (only software) in bootstrap.c

**Comment by tjaychen** [comment]
is there any way you could push this to a fork and i can pull it to have a
look?

also, are you having issues specifically to compression with bootstrap
calls? I'm wondering if we should just debug separately if the compression
call is causing the core to die.

when you just increase the ROM size (without any change in function), does
everything still work okay?

**Comment by Avani8** [comment]
@tjaychen here is the [link](https://github.com/Avani8/otcompression) for the fork. 
not sure the issue is for only compression function, I think if the code size is increasing 1k or something it fails hash matching .. for example in the fork look at sw/device/lib/hello.c where I have sum/sub/mul/div func along with compression code. I have validated that sum works with ROM size increase but when I include compression func (API ) in bootstrap.c it gets stuck at frame 1.

**Comment by tjaychen** [comment]
Thanks!
I still have not had a chance to look into this yet, but I will try to find
time later this week.

**Comment by Avani8** [comment]
@tjaychen did you get a chance to look at this? I am still stuck. I think there is something blocking ROM size extension.  as I tried to extend ROM to 0x7000 and had approx 1K data to dump on the memory I am getting following error [
![Screenshot from 2020-01-17 16-17-56](https://user-images.githubusercontent.com/12041484/72646923-40264d00-3945-11ea-8b6b-899eb2eb53b9.png)
](url)

**Comment by tjaychen** [comment]
sorry @Avani8 i got kind of tied up looking at other things this week.

I did just pull your patch down to try however..and I was not able to reproduce your latest issue.  Using ROM_size extension 0x7000, and building with your version of ROM (i see the lz77_compress code in bootstrap.c and hello.c) I did not get the issue you are showing above.

What version of of the riscv-toolchain do you have installed? Mine looks to be 8.3.0

chip_info located at 0x9xxx should be okay i think, since our ROM base address is 0x8000 and that should be plenty enough space. 

I don't have access to an FPGA right now, trying to see if i can reproduce your error on verilator..

**Comment by tjaychen** [comment]
one other thing that occurs to me... it looks like the compress / uncompress is in the bootstrap function. Do you know have any rough idea how long these functions take? The spi transaction signaling isn't super robust right now, and relies on some hardcoded timing, basically assuming the device side can finish hashing and send back the results within a certain time.  I am not sure what placing a big activity in the middle of that might do. 

I'm guessing you've tried smaller functions in there, and everything worked out just fine?

**Comment by Avani8** [comment]
for reproducing today's error: I've created uint32_t txt[984] = {somevalues} and wrote a for loop to print the txt[i] 32 bit each using uart_send_int(). 

Since we had an issue with extending memory size we have changed our work direction to have either compression/ decompression function support, not both. so approx we need 3-4 kb extension in ROM size 

yes for smaller function everything works.

also thanks for base address pointer, 
base address: 0x8000 and size is: 0x2000, which includes chip_info located at 0x9xxx . I've changed starting address to support at least 4k + memory area and running a test on FPGA ...will share the results in few

**Comment by tjaychen** [comment]
I still think there might be issues if you do that within bootstrap. Could
you move your test code to the beginning of bootrom and leave bootstrap as
is? I really think the timing might be affected.

**Comment by Avani8** [comment]
moving code to boot_rom instead of bootstrap defeats the purpose. what we are trying to do here is when hash matching fails we will have this decompressor trigger and get a fresh copy of the data[] and replace it with the affected code. this is a proof of concept work which we are validating in our research paper.

**Comment by tjaychen** [comment]
Yes I understand there's no point, but I think we are probably messing some
of the assumed timing between spiflash and ROM.

What you could also try is go into the spiflash code and update the timing
so that the host waits longer for each frame. I'm asking for the move
mostly just to confirm if it's a timing issue or something more
fundamental.

**Comment by Avani8** [comment]
got it trying now (just for a tet run) to see the timing issue as you suggested. will post results in few. 

btw got the result of the previous run with just changing base address: 4000 instead of  0x8000. with the size : 4000 instead of 2000.  it fails at frame one check as previously. 
did run another test with just changing ROM size: 0x4000 and start : 0x6000 (16kb) without changing anything in bootstrap.c (original code from opentitan) . but still, it fails at frame one  check  
does this mean I cannot change the starting address 
![Screenshot from 2020-01-17 19-50-11](https://user-images.githubusercontent.com/12041484/72655456-b38a8780-3962-11ea-9f80-775e4e1c35f0.png)

**Comment by tjaychen** [comment]
Could you show the uart output when you change the base..? I don't remember
anything that restricts it, but I would have to look around.

**Comment by Avani8** [comment]
it's as attached screen capture 
![Screenshot from 2020-01-17 19-50-11](https://user-images.githubusercontent.com/12041484/72656114-357caf80-3967-11ea-9534-351dc0c6d6c3.png)

**Comment by Avani8** [comment]
This means I can change the ROM size but not the starting address

**Comment by tjaychen** [comment]
Okay one thing I completely forgot. When you change the ROM size you
probably need to alter the rom_link.ld file under the boot ROM directory.

That makes an assumption regarding the size of the ROM and places the chip
info stuff at a specific place (probably not necessary, but I'd want to
check with mcy for that).

Change that to match your actual size. So for your case, don't change the
base for now, just enlarge the size to accommodate your needs. That should
move the chip info stuff also.

**Comment by Avani8** [comment]
ok trying now

**Comment by Avani8** [comment]
ahh...finally...got it fixed with changing in .hjson and .ld places the size...thanks a lot @tjaychen for all help and troubleshooting. I will put your name in credits for the paper. will reach out to you with the draft. thanks again.

**Comment by imphil** [comment]
Looks like this is resolved, please reopen if that's not the case.
